wpengine/atlas-content-modeler
# [BH-1147] Format Model IDs as WordPress post type slugs

## Description


The primary goal of this PR is to standardize how we format model slugs (also known as Model IDs). All model slugs eventually become WordPress post type slugs. Using the same formatting rules for model slugs that WP uses for post types should eliminate or reduce the need to reformat model slugs before using them to reference post types.

#### The REST routes for creating/updating content models now format model slugs with `sanitize_key()` before saving.

This is primarily handled through updates to the `create_model()` function. It's also worth noting that the `create_model()` function now returns the created model rather than a boolean so that clients know if `sanitize_key()` has resulted in a different model slug than what was submitted.

Model update routes did not require any special consideration because changing a model slug is currently unsupported. However, there is a small, one line change that makes sure this limitation is enforced.

#### The DevEx application formats model slugs to match the format defined by the WordPress `sanitize_key()` function.

This was accomplished by refactoring the `useApiIdGenerator` hook to be more generic. The hook is now named `useInputGenerator` and allows the client to supply their own formatting function. As a result, we can use `toValidApiId()` to generate field slugs and `toPostTypeSlug()` to generate model slugs.

#### Any references to model slugs have been updated to account for the new format.

- Several cases of lowercasing a model slug before it is used have been removed.
- Model and Field REST routes that include a model slug have an updated regex that matches `sanitize_key()` style slugs.

#### Backwards Compatibility

It's fully backwards compatible! We were consistently using a wrapper function around the `get_option()` call for retrieving models, so that was a natural place to reformat existing model slugs. ~~Since all of our models are stored together in a single option, anything that triggers an option write in the plugin (i.e. creating new models, updating or adding fields to models) will automatically correct all of the unformatted slugs in the DB.~~ We are now persisting updated model slugs on retrieval if any of the existing slugs are changed. I've included a TODO note to consider removing this compatibility patch when we move to v1.0 so that we're not stuck with it forever.

I also explored the idea of including a very simple migration system, but scrapped it in favor of the lazy update approach. We'll likely want to revisit the idea at some point, but such a task felt like it deserved its own separate PR and suite of tests.



https://wpengine.atlassian.net/browse/BH-1147

## Testing



- PHP unit tests are included.
- Unit test for the `toPostTypeSlug()` formatter included.
- I had trouble getting Jest tests written for the `useInputGenerator` hook, but I think a majority of that functionality is covered in existing e2e tests.

#### Manual testing steps focused on backwards compatibility

1. Before checking out this branch, create the following three models entering only the singular/plural labels and taking the defaults for everything else:
    1. Rabbit/Rabbits
    2. Two Word/Two Words
    3. Special-Character_9/Special-Characters_9
2. Check out this branch and rebuild the plugin.
3. Refresh the WordPress dashboard and see that all three models you created still exist.
4. Look at the `atlas_content_modeler_post_types` option in your database and verify that all models slugs have been lowercased (or more specifically, contain only lowercase alphanumeric characters, dashes, and underscores)
5. Verify that model creation, model editing, and model field creation still works.

## Dependant PRs


The `taxonomies` branch will need to be updated once this has been merged. I will take care of that reconciliation.
